internal/engine/upper.go

```diff
@@ -25,7 +35,34 @@ func NewUpper(manager service.Manager) (Upper, error) {
 	watcherMaker := func() (watch.Notify, error) {
 		return watch.NewWatcher()
 	}
-	return Upper{b, watcherMaker}, nil
+	sleeper := time.Sleep
+	return Upper{b, watcherMaker, sleeper}, nil
+}
+
+//makes an attempt to read some events from `eventChan` so that multiple file changes that happen at the same time
+//from the user's perspective are grouped together.
+func (u Upper) coalesceEvents(eventChan chan fsnotify.Event) []fsnotify.Event {
+	var events []fsnotify.Event
+	doneWaitingForChanges := false
+	for !doneWaitingForChanges {
```

Comment:
the double loop here seems overly complex. One of the great powers of `select` is to avoid this kind of nested-loops-to-handle-multiple-conditions

this would be a lot simpler as

```
timer := time.NewTimer(delay)
for true {
  select {
    case event := <- eventChan:
      timer.Reset(delay)
       ...
    case <-timer.C:
       ...
  }
}
```

Reply:
yeah, that's much better. thanks.